frmts/sentinel2/sentinel2dataset.cpp

```diff
@@ -438,6 +440,33 @@ int SENTINEL2Dataset::Identify( GDALOpenInfo *poOpenInfo )
         strstr(pszHeader, "User_Product_Level-2A" ) != nullptr )
         return TRUE;
 
+    /* Last effort – look inside of a zip file for metadata file with
+     * matching name. */
+    if ( EQUAL(CPLGetExtension(pszJustFilename), "zip") )
+    {
+        CPLString osFilename(poOpenInfo->pszFilename);
+        if( strncmp(osFilename, "/vsizip/", strlen("/vsizip/")) != 0 )
+            osFilename = "/vsizip/" + osFilename;
+
+        auto psDir = VSIOpenDir(osFilename.c_str(), 1, nullptr);
```

Comment:
psDir must be freed with VSICloseDir(). Same in the Open() method

Reply:
Thanks!